File: Python/4_1_4_TI_Optimization/main_ti.py
#定量的マクロ経済学と数値計算 第4章 オイラー方程式と多項式近似
#時間反復法: メインファイル

#必要な関数・モジュールを呼び出す
import numpy as np
from scipy.interpolate import interp1d
import matplotlib.pyplot as plt
import japanize_matplotlib
from CRRA import CRRA, mu_CRRA
from EulerEq import EulerEq
from nti import nti

#カリブレーション
class Model():
    """
    モデルを解くためのパラメータを含む class を定義する。
    """
    def __init__(self,
        beta = 0.96,    # 割引因子
        gamma = 1.0,    # 相対的リスク回避度(異時点間の代替弾力性の逆数)
        alpha = 0.4,    # 資本分配率 
        delta = 1.0,    # 固定資本減耗
        nk = 21,        # 資本のグリッド数
        kmax = 0.5,     # 資本グリッドの最大値
        kmin = 0.05,    # 資本グリッドの最小値
        maxiter = 1000, # 繰り返し計算の最大値
        tol = 1e-5,     # 許容誤差(STEP2)
        ): 
        
        self.beta, self.gamma, self.alpha = beta, gamma, alpha 
        self.delta, self.nk = delta, nk 
        self.kmax, self.kmin = kmax, kmin 
        self.kgrid = np.linspace(kmin,kmax,nk) # 資本のグリッド
        self.maxiter, self.tol = maxiter, tol

print("")
print("-+- Solve a neoclassical growth model with time iteration -+-")
print("")

m = Model()
mu_CRRA_vec = np.vectorize(mu_CRRA)

#時間反復法を実施
cfcn0 = nti(m)

#最終的な政策関数が得られてから貯蓄関数を計算
pfcn0 =  (m.kgrid ** m.alpha) + (1-m.delta)*m.kgrid - cfcn0 

#解析的解(for k'=g(k))
p_true = m.beta * m.alpha * (m.kgrid ** m.alpha)

#オイラー方程式から誤差を測定
kgrid_err = np.linspace(m.kmin,m.kmax,(m.nk-1)*10+1)
cons_interp = interp1d(m.kgrid,cfcn0,kind="linear",fill_value="extrapolate")
cons = cons_interp(kgrid_err)
LHS = mu_CRRA_vec(m,cons)

kp = (kgrid_err ** m.alpha) + (1-m.delta)*kgrid_err - cons
cnext = cons_interp(kp)
rent = m.alpha*(kp ** (m.alpha-1)) - m.delta
RHS = m.beta * (1+rent) * mu_CRRA_vec(m,cnext)

err = (RHS/LHS) - 1

#図を描く
fig, ax = plt.subplots()
ax.plot(m.kgrid,pfcn0,label="近似解")
ax.plot(m.kgrid,p_true,ls="--",label="解析的解")
ax.plot(m.kgrid,m.kgrid,label="45度線")
ax.set(xlabel=r"今期の資本保有量: $k$",ylabel=r"次期の資本保有量: $k'$",title=r"(a):政策関数:$g(x)=f(x)-h(x)$",
xlim=(m.kmin,m.kmax),xticks=[0.05,0.1,0.2,0.3,0.4,0.5])
ax.legend(loc="upper left")
ax.grid()
plt.show()

f = open("err_ndp.csv",encoding='utf-8-sig')
err2 = np.loadtxt(f) #VFIでのオイラー方程式誤差

fig, ax = plt.subplots()
ax.plot(kgrid_err,np.abs(err),label="TI")
ax.plot(kgrid_err,np.abs(err2),ls="--",label="VFI")
ax.set(xlabel=r"資本保有量: $k$",ylabel="オイラー方程式誤差(絶対値)",title="(b):オイラー方程式に代入した際の誤差",
xlim=(m.kmin,m.kmax),xticks=[0.05,0.1,0.2,0.3,0.4,0.5])
ax.legend(loc="upper right")
ax.grid()
plt.show()

#TIとVFIでのオイラー方程式誤差を比較する
L1_TI = np.log10(np.mean(np.abs(err)))
Lmax_TI = np.log10(np.max(np.abs(err)))

L1_VFI = np.log10(np.mean(np.abs(err2)))
Lmax_VFI = np.log10(np.max(np.abs(err2)))

print("")
print("method", "[log10(L1), log10(L∞)]")
print("TI", [L1_TI, Lmax_TI])
print("VFI", [L1_VFI, Lmax_VFI])
print("")

File: Matlab/4_1_4_TI_Optimization/main_ti.m
clear all;

% MATLABではグローバル変数を使って関数への変数受け渡しを行う
% Julia, Pythondではコンストラクタを用いる
global beta gamma alpha delta kgrid

% カリブレーション
beta  = 0.96; % 割引因子
gamma = 1.0;  % 相対的危険回避度(異時点間の代替の弾力性の逆数)
alpha = 0.40; % 資本分配率
delta = 1.00; % 固定資本減耗(delta=1.0のときは解析解が存在)

% 定常状態の値
ykss = (1/beta-1+delta)/alpha;
kss = ykss^(1/(alpha-1));
yss = ykss*kss;
css = yss-delta*kss;

kmax = 0.5;   % 資本グリッドの最大値
kmin = 0.05;  % 資本グリッドの最小値 (0にすると生産が出来なくなる)

%% STEP 1(a): グリッド生成
nk   = 21;    % グリッドの数
kgrid = linspace(kmin, kmax, nk)';

maxiter = 1000; % 繰り返し計算の最大値
tol  = 1.0e-5;  % 許容誤差(STEP 2)

% 収束の基準に関するパラメータ
it = 1;          % ループ・カウンター
dif2 = 1.0;      % 政策関数の繰り返し誤差
options.TolFun = 1.0e-10; % fsolveのオプション(最適化の許容誤差)

tic;
options = optimoptions('fsolve','Display','none'); % fsolveのオプション(最適化の結果を非表示にする)

disp(' ')
disp('-+- Solve a neoclassical growth model with time iteration -+-');
disp(' ')

%% STEP 1(b): 政策関数の初期値を当て推量
cfcn0 = kgrid;
cfcn1 = zeros(nk,1);

% 繰り返し誤差を保存する変数を設定 
dif = zeros(2,maxiter);

%% STEP 4: 政策関数を繰り返し計算
while (it < maxiter && dif2 > tol)

    for i = 1:nk

        capital = kgrid(i);
        wealth = capital.^alpha + (1.-delta).*capital;

        % MATLABの最適化関数(fsolve)を使って各グリッド上の政策関数の値を探す
        % 最適化の初期値は古い政策関数の値
        cons = fsolve(@EulerEq,cfcn0(i,1),options,capital,cfcn0);
        cfcn1(i,1) = cons;
        kprime = wealth-cons;
        % グリッドごとに最適化の結果を確認
        %disp([cons capital wealth kprime]);
        %pause

    end

    % 繰り返し計算誤差を確認
    dif2 = max(abs(cfcn1-cfcn0));

    % 収束途中の繰り返し計算誤差を保存
    dif(2,it) = dif2;
    
    % 政策関数をアップデート
    cfcn0 = cfcn1;

    fprintf('iteration index: %i \n', it);
    fprintf('policy function iteration error: %1.6f \n', dif2);

    it = it + 1;

end

disp(' ');
toc;

%% 最終的な政策関数が得られてから貯蓄関数を計算
pfcn0 = kgrid.^alpha + (1-delta)*kgrid - cfcn0;

%% 解析的解
p_true = beta*alpha*(kgrid.^alpha);

%%
figure;
plot(kgrid, pfcn0, '-', 'Color', 'blue', 'LineWidth', 3);
hold on;
plot(kgrid, p_true, '--', 'Color', 'red', 'LineWidth', 3);
plot(kgrid, kgrid, ':', 'Color', 'black', 'LineWidth', 2);
xlabel('今期の資本保有量：k', 'FontSize', 16);
ylabel("次期の資本保有量：k'", 'FontSize', 16);
xlim([kmin kmax]);
xticks([0.05 0.1 0.2 0.3 0.4 0.5]);
xticklabels([0.05 0.1 0.2 0.3 0.4 0.5]);
legend('近似解', '解析的解', '45度線', 'Location', 'NorthWest');
grid on;
set(gca,'FontSize', 16);
saveas(gcf,'Fig_pti2.eps','epsc2');

%% オイラー方程式から誤差を測定
% 元のグリッドではオイラー方程式の誤差はゼロになるため、グリッドを細かくとる
kgrid_err = linspace(kmin, kmax, (nk-1)*10+1)';
% c=h(k;b)の値を線形補間
cons = interp1(kgrid,cfcn0(:,1),kgrid_err);
LHS  = mu_CRRA(cons, gamma);

% k'=f(k)-c
kp   = kgrid_err.^alpha + (1-delta)*kgrid_err - cons;
% c'=h(k';b)の値を線形補間
cnext = interp1(kgrid, cfcn0(:,1), kp);
rent = alpha.*kp.^(alpha-1.0) - delta;
RHS  = beta.*(1.+rent).*mu_CRRA(cnext,gamma);

err  = RHS./LHS-1.0;

err2 = csvread("err_ndp.csv");
figure;
plot(kgrid_err, abs(err), '-', 'Color', 'blue', 'LineWidth', 3);
hold on;
plot(kgrid_err, abs(err2), '--', 'Color', 'red', 'LineWidth', 3);
xlabel('資本保有量：k', 'FontSize', 16);
ylabel('オイラー方程式誤差(絶対値)', 'FontSize', 16);
xlim([kmin kmax]);
xticks([0.05 0.1 0.2 0.3 0.4 0.5]);
xticklabels([0.05 0.1 0.2 0.3 0.4 0.5]);
legend('TI', 'VFI', 'Location', 'NorthEast');
grid on;
set(gca,'FontSize', 16);
saveas (gcf,'Fig_pti6.eps','epsc2');
